File: app/schemas/usage.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.usage import (
    AccountingStatus,
    UsageChargeStatus,
    UsageRatingRunStatus,
    UsageSource,
)


class QuotaBucketBase(BaseModel):
    subscription_id: UUID
    period_start: datetime
    period_end: datetime
    included_gb: Decimal | None = None
    used_gb: Decimal = Decimal("0")
    rollover_gb: Decimal = Decimal("0")
    topup_gb: Decimal = Decimal("0")
    overage_gb: Decimal = Decimal("0")


class QuotaBucketCreate(QuotaBucketBase):
    pass


class QuotaBucketUpdate(BaseModel):
    subscription_id: UUID | None = None
    period_start: datetime | None = None
    period_end: datetime | None = None
    included_gb: Decimal | None = None
    used_gb: Decimal | None = None
    rollover_gb: Decimal | None = None
    topup_gb: Decimal | None = None
    overage_gb: Decimal | None = None


class QuotaBucketRead(QuotaBucketBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    created_at: datetime
    updated_at: datetime
    # Running cost of the current overage (overage_gb × the allowance's
    # overage_rate). Populated on customer-facing reads; None when not in
    # overage or no rate is configured.
    overage_amount: Decimal | None = None


class RadiusAccountingSessionBase(BaseModel):
    subscription_id: UUID
    access_credential_id: UUID
    radius_client_id: UUID | None = None
    nas_device_id: UUID | None = None
    session_id: str = Field(min_length=1, max_length=120)
    status_type: AccountingStatus
    session_start: datetime | None = None
    session_end: datetime | None = None
    # Most recent accounting observation (interim update or stop); a live
    # session keeps advancing this, an open one gone quiet is a ghost.
    last_update_at: datetime | None = None
    input_octets: int | None = None
    output_octets: int | None = None
    terminate_cause: str | None = Field(default=None, max_length=120)
    framed_ip_address: str | None = Field(default=None, max_length=64)
    framed_ipv6_prefix: str | None = Field(default=None, max_length=128)
    delegated_ipv6_prefix: str | None = Field(default=None, max_length=128)
    nas_port_id: str | None = Field(default=None, max_length=64)
    called_station_id: str | None = Field(default=None, max_length=64)


class RadiusAccountingSessionCreate(RadiusAccountingSessionBase):
    calling_station_id: str | None = Field(default=None, max_length=64)


class RadiusAccountingSessionUpdate(BaseModel):
    subscription_id: UUID | None = None
    access_credential_id: UUID | None = None
    radius_client_id: UUID | None = None
    nas_device_id: UUID | None = None
    session_id: str | None = Field(default=None, max_length=120)
    status_type: AccountingStatus | None = None
    session_start: datetime | None = None
    session_end: datetime | None = None
    last_update_at: datetime | None = None
    input_octets: int | None = None
    output_octets: int | None = None
    terminate_cause: str | None = Field(default=None, max_length=120)
    framed_ip_address: str | None = Field(default=None, max_length=64)
    framed_ipv6_prefix: str | None = Field(default=None, max_length=128)
    delegated_ipv6_prefix: str | None = Field(default=None, max_length=128)
    nas_port_id: str | None = Field(default=None, max_length=64)
    called_station_id: str | None = Field(default=None, max_length=64)
    calling_station_id: str | None = Field(default=None, max_length=64)


class RadiusAccountingSessionRead(RadiusAccountingSessionBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    created_at: datetime


class UsageSeriesPoint(BaseModel):
    bucket_start: datetime
    bytes: int


class FupSummary(BaseModel):
    """Customer-facing Fair-Usage status for the caller's subscription.

    Derived from the per-subscription ``FupState`` the enforcement engine
    maintains; surfaced so the app can tell a subscriber they're throttled,
    how to restore speed, and when the limit resets.
    """

    status: str = Field(description="full_speed | approaching | throttled | blocked")
    is_reduced: bool = False
    speed_reduction_percent: float | None = None
    active_rule_name: str | None = None
    resets_at: datetime | None = None
    # Plain-language explainer for the active rule, e.g.
    # "Speed reduced to 25% after 100 GB this month".
    summary: str | None = None
    # Headroom against the nearest throttle/block rule — present even while
    # healthy so the app can pre-warn before enforcement.
    threshold_gb: float | None = None
    used_gb: float | None = None
    gb_until_throttle: float | None = None
    usage_ratio: float | None = None
    # Policy terms shown regardless of state, e.g.
    # "Speed reduces to 25% after 500 GB each month".
    policy_summary: str | None = None


class UsageSummaryResponse(BaseModel):
    """Time-windowed data-usage summary for GET /me/usage-summary."""

    period: str = Field(description="hour | today | week | cycle | all")
    start: datetime
    end: datetime
    total_bytes: int
    # Where the headline total came from: "samples" (integrated throughput),
    # "sessions" (RADIUS accounting octets), or "quota" (rated billing usage).
    total_source: str
    # True when the total is billing-grade (quota / session octets) rather than
    # reconstructed from the throughput series.
    is_authoritative: bool
    # Bucket width of the series: "minute" | "hour" | "day" | None (no chart).
    bucket: str | None = None
    # Mean throughput over the window (rx+tx bits/s) — the "average speed".
    # None for windows with no sample points (e.g. "all").
    average_bps: float | None = None
    series: list[UsageSeriesPoint] = Field(default_factory=list)
    # Fair-Usage status for the caller (None when no FUP applies / unknown).
    fup: FupSummary | None = None


class DailyUsagePoint(BaseModel):
    """One day's upload/download volume (bytes), summed across the caller's
    subscriptions."""

    date: date
    upload_bytes: int
    download_bytes: int
    total_bytes: int


class DailyUsageHistoryResponse(BaseModel):
    """Long-history daily usage for GET /me/usage-history.

    Sourced from the historical daily rollup (Splynx ``traffic_counter``
    backfill), which reaches years further back than per-session accounting.
    """

    start: date
    end: date
    total_upload_bytes: int
    total_download_bytes: int
    total_bytes: int
    points: list[DailyUsagePoint] = Field(default_factory=list)


class UsageRecordBase(BaseModel):
    subscription_id: UUID
    quota_bucket_id: UUID | None = None
    source: UsageSource
    recorded_at: datetime
    input_gb: Decimal = Decimal("0")
    output_gb: Decimal = Decimal("0")
    total_gb: Decimal = Decimal("0")


class UsageRecordCreate(UsageRecordBase):
    pass


class UsageRecordUpdate(BaseModel):
    subscription_id: UUID | None = None
    quota_bucket_id: UUID | None = None
    source: UsageSource | None = None
    recorded_at: datetime | None = None
    input_gb: Decimal | None = None
    output_gb: Decimal | None = None
    total_gb: Decimal | None = None


class UsageRecordRead(UsageRecordBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    created_at: datetime


class UsageChargeBase(BaseModel):
    subscription_id: UUID
    subscriber_id: UUID = Field(
        validation_alias="account_id", serialization_alias="account_id"
    )
    invoice_line_id: UUID | None = None
    period_start: datetime
    period_end: datetime
    total_gb: Decimal = Field(default=Decimal("0.0000"), ge=0)
    included_gb: Decimal = Field(default=Decimal("0.0000"), ge=0)
    billable_gb: Decimal = Field(default=Decimal("0.0000"), ge=0)
    unit_price: Decimal = Field(default=Decimal("0.0000"), ge=0)
    amount: Decimal = Field(default=Decimal("0.00"), ge=0)
    currency: str = Field(default="NGN", min_length=3, max_length=3)
    status: UsageChargeStatus = UsageChargeStatus.staged
    notes: str | None = None
    rated_at: datetime | None = None


class UsageChargeRead(UsageChargeBase):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    created_at: datetime
    updated_at: datetime


class UsageRatingRunRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    run_at: datetime
    period_start: datetime
    period_end: datetime
    status: UsageRatingRunStatus
    subscriptions_scanned: int
    charges_created: int
    skipped: int
    error: str | None
    created_at: datetime


class UsageRatingRunRequest(BaseModel):
    period_start: datetime | None = None
    period_end: datetime | None = None
    subscription_id: UUID | None = None
    dry_run: bool = False


class UsageRatingRunResponse(BaseModel):
    run_id: UUID | None = None
    run_at: datetime
    period_start: datetime
    period_end: datetime
    subscriptions_scanned: int
    charges_created: int
    skipped: int


class UsageChargePostRequest(BaseModel):
    invoice_id: UUID | None = None


class UsageChargePostBatchRequest(BaseModel):
    period_start: datetime
    period_end: datetime
    account_id: UUID | None = None


class UsageChargePostBatchResponse(BaseModel):
    posted: int

File: mobile/lib/src/models/usage.dart
// Mirrors QuotaBucketRead from app/schemas/usage.py.
import '../core/parsers.dart';

class QuotaBucket {
  QuotaBucket({
    required this.id,
    required this.subscriptionId,
    required this.periodStart,
    required this.periodEnd,
    this.includedGb,
    this.usedGb = 0,
    this.rolloverGb = 0,
    this.topupGb = 0,
    this.overageGb = 0,
    this.overageAmount,
  });

  final String id;
  final String subscriptionId;
  final DateTime periodStart;
  final DateTime periodEnd;
  final double? includedGb;
  final double usedGb;
  final double rolloverGb;
  final double topupGb;
  final double overageGb;

  /// Running cost of the current overage (₦), when the plan has an overage
  /// rate. Null when not in overage or unmetered.
  final double? overageAmount;

  /// Total data available this period (included + rolled-over + top-ups).
  double? get allowanceGb =>
      isUnlimited ? null : includedGb! + rolloverGb + topupGb;

  double? get remainingGb {
    final a = allowanceGb;
    if (a == null) return null;
    final r = a - usedGb;
    return r < 0 ? 0 : r;
  }

  /// 0..1 fraction of the allowance consumed, or null for unlimited plans.
  double? get usedFraction {
    final a = allowanceGb;
    if (a == null || a <= 0) return null;
    final f = usedGb / a;
    return f.clamp(0.0, 1.0);
  }

  /// Unmetered plan: no included allowance on file (null) or an explicit 0,
  /// which is how "unlimited" offers are rated (no GB cap to count against).
  bool get isUnlimited => includedGb == null || includedGb! <= 0;

  factory QuotaBucket.fromJson(Map<String, dynamic> json) => QuotaBucket(
        id: json['id'].toString(),
        subscriptionId: json['subscription_id'].toString(),
        periodStart: DateTime.parse(json['period_start'].toString()).toLocal(),
        periodEnd: DateTime.parse(json['period_end'].toString()).toLocal(),
        includedGb: asDoubleOrNull(json['included_gb']),
        usedGb: asDouble(json['used_gb']),
        rolloverGb: asDouble(json['rollover_gb']),
        topupGb: asDouble(json['topup_gb']),
        overageGb: asDouble(json['overage_gb']),
        overageAmount: asDoubleOrNull(json['overage_amount']),
      );
}

/// Mirrors RadiusAccountingSessionRead from app/schemas/usage.py.
class AccountingSession {
  AccountingSession({
    required this.id,
    required this.subscriptionId,
    required this.sessionId,
    required this.statusType,
    this.sessionStart,
    this.sessionEnd,
    this.lastUpdateAt,
    this.inputOctets,
    this.outputOctets,
    this.terminateCause,
    this.framedIpAddress,
  });

  final String id;
  final String subscriptionId;
  final String sessionId;
  final String statusType;
  final DateTime? sessionStart;
  final DateTime? sessionEnd;

  /// Most recent accounting observation (interim update or stop). For an
  /// active session this is when it was last seen alive — connect time alone
  /// can be days old on a healthy always-on PPPoE session.
  final DateTime? lastUpdateAt;
  final int? inputOctets;
  final int? outputOctets;
  final String? terminateCause;

  /// The IP the NAS assigned this session (RADIUS Framed-IP-Address) — the
  /// subscriber's live address, present even for dynamically-addressed plans.
  final String? framedIpAddress;

  /// Total bytes transferred (down + up). RADIUS input = from the NAS toward
  /// the subscriber's perspective varies by vendor; we just sum both.
  int get totalOctets => (inputOctets ?? 0) + (outputOctets ?? 0);

  bool get isActive => sessionEnd == null;

  DateTime? get lastSeenAt => lastUpdateAt ?? sessionEnd ?? sessionStart;

  factory AccountingSession.fromJson(Map<String, dynamic> json) =>
      AccountingSession(
        id: json['id'].toString(),
        subscriptionId: json['subscription_id'].toString(),
        sessionId: json['session_id'].toString(),
        statusType: json['status_type'].toString(),
        sessionStart: _toDate(json['session_start']),
        sessionEnd: _toDate(json['session_end']),
        lastUpdateAt: _toDate(json['last_update_at']),
        inputOctets: (json['input_octets'] as num?)?.toInt(),
        outputOctets: (json['output_octets'] as num?)?.toInt(),
        terminateCause: json['terminate_cause'] as String?,
        framedIpAddress: json['framed_ip_address'] as String?,
      );
}

/// Live throughput for the customer's active subscription, subscriber
/// perspective. Mirrors BandwidthStats from api/bandwidth.py (we bind only
/// the download/upload fields — rx/tx are NAS-perspective).
class LiveBandwidth {
  LiveBandwidth({
    this.downloadBps,
    this.uploadBps,
    this.peakDownloadBps,
    this.peakUploadBps,
  });

  final double? downloadBps;
  final double? uploadBps;

  /// Peak throughput over the requested window (subscriber perspective).
  /// Populated by /bandwidth/my/stats; null on the live one-shot.
  final double? peakDownloadBps;
  final double? peakUploadBps;

  bool get hasSignal => (downloadBps ?? 0) > 0 || (uploadBps ?? 0) > 0;

  factory LiveBandwidth.fromJson(Map<String, dynamic> json) => LiveBandwidth(
        downloadBps: (json['download_bps'] as num?)?.toDouble(),
        uploadBps: (json['upload_bps'] as num?)?.toDouble(),
        peakDownloadBps: (json['peak_download_bps'] as num?)?.toDouble(),
        peakUploadBps: (json['peak_upload_bps'] as num?)?.toDouble(),
      );
}

/// One point of the bandwidth-speed time series. Mirrors BandwidthSeriesPoint
/// from app/api/bandwidth.py (GET /bandwidth/my/series). We bind only the
/// subscriber-perspective download/upload rates.
class BandwidthPoint {
  BandwidthPoint({
    required this.at,
    required this.downloadBps,
    required this.uploadBps,
  });

  final DateTime at;
  final double downloadBps;
  final double uploadBps;

  double get totalBps => downloadBps + uploadBps;

  factory BandwidthPoint.fromJson(Map<String, dynamic> json) => BandwidthPoint(
        at: DateTime.parse(json['timestamp'].toString()).toLocal(),
        downloadBps: (json['download_bps'] as num?)?.toDouble() ?? 0,
        uploadBps: (json['upload_bps'] as num?)?.toDouble() ?? 0,
      );
}

/// One bar of the usage chart. Mirrors UsageSeriesPoint from schemas/usage.py.
class UsageSeriesPoint {
  UsageSeriesPoint({required this.bucketStart, required this.bytes});

  final DateTime bucketStart;
  final int bytes;

  factory UsageSeriesPoint.fromJson(Map<String, dynamic> json) =>
      UsageSeriesPoint(
        bucketStart: DateTime.parse(json['bucket_start'].toString()).toLocal(),
        bytes: (json['bytes'] as num?)?.toInt() ?? 0,
      );
}

/// Customer-facing Fair-Usage status. Mirrors FupSummary in schemas/usage.py.
class FupStatus {
  FupStatus({
    required this.status,
    this.isReduced = false,
    this.speedReductionPercent,
    this.activeRuleName,
    this.resetsAt,
    this.summary,
    this.thresholdGb,
    this.usedGb,
    this.gbUntilThrottle,
    this.usageRatio,
    this.policySummary,
  });

  final String status; // full_speed | approaching | throttled | blocked
  final bool isReduced;
  final double? speedReductionPercent;
  final String? activeRuleName;
  final DateTime? resetsAt;
  final String? summary; // plain-language explainer

  /// Headroom against the nearest throttle/block rule — present even while
  /// healthy so the app can pre-warn before enforcement.
  final double? thresholdGb;
  final double? usedGb;
  final double? gbUntilThrottle;
  final double? usageRatio;

  /// Policy terms shown regardless of state, e.g.
  /// "Speed reduces to 25% after 500 GB each month".
  final String? policySummary;

  bool get isThrottled => status == 'throttled';
  bool get isBlocked => status == 'blocked';
  bool get isApproaching => status == 'approaching';

  /// Whether the customer should see a banner / restore CTA.
  bool get needsAttention => isThrottled || isBlocked;

  factory FupStatus.fromJson(Map<String, dynamic> json) => FupStatus(
        status: json['status']?.toString() ?? 'full_speed',
        isReduced: json['is_reduced'] as bool? ?? false,
        speedReductionPercent:
            (json['speed_reduction_percent'] as num?)?.toDouble(),
        activeRuleName: json['active_rule_name'] as String?,
        resetsAt: _toDate(json['resets_at']),
        summary: json['summary'] as String?,
        thresholdGb: (json['threshold_gb'] as num?)?.toDouble(),
        usedGb: (json['used_gb'] as num?)?.toDouble(),
        gbUntilThrottle: (json['gb_until_throttle'] as num?)?.toDouble(),
        usageRatio: (json['usage_ratio'] as num?)?.toDouble(),
        policySummary: json['policy_summary'] as String?,
      );
}

/// Windowed data-usage summary. Mirrors UsageSummaryResponse from
/// schemas/usage.py (GET /me/usage-summary).
class UsageSummary {
  UsageSummary({
    required this.period,
    required this.start,
    required this.end,
    required this.totalBytes,
    required this.totalSource,
    required this.isAuthoritative,
    this.bucket,
    this.averageBps,
    this.peakDownloadBps,
    this.peakUploadBps,
    this.series = const [],
    this.fup,
  });

  final String period; // hour | today | week | cycle | all
  final DateTime start;
  final DateTime end;
  final int totalBytes;
  final String totalSource; // samples | sessions | quota | lifetime
  final bool isAuthoritative;

  /// Billing-grade headline total when the server says this window is
  /// authoritative. Zero is a valid value; null means the loaded summary is
  /// estimated and must not be replaced with a client-side reconstruction.
  int? get authoritativeTotalBytes => isAuthoritative ? totalBytes : null;

  final String? bucket; // minute | hour | day | null

  /// Mean throughput over the window (rx+tx bits/s) — the "average speed".
  /// Null for windows with no samples (e.g. "all").
  final double? averageBps;

  /// Exact peak throughput over the window (subscriber bits/s). Populated for
  /// the billing cycle; null when unavailable.
  final double? peakDownloadBps;
  final double? peakUploadBps;
  final List<UsageSeriesPoint> series;
  final FupStatus? fup;

  factory UsageSummary.fromJson(Map<String, dynamic> json) => UsageSummary(
        period: json['period'].toString(),
        start: DateTime.parse(json['start'].toString()).toLocal(),
        end: DateTime.parse(json['end'].toString()).toLocal(),
        totalBytes: (json['total_bytes'] as num?)?.toInt() ?? 0,
        totalSource: json['total_source'].toString(),
        isAuthoritative: json['is_authoritative'] as bool? ?? false,
        bucket: json['bucket'] as String?,
        averageBps: (json['average_bps'] as num?)?.toDouble(),
        peakDownloadBps: (json['peak_download_bps'] as num?)?.toDouble(),
        peakUploadBps: (json['peak_upload_bps'] as num?)?.toDouble(),
        series: (json['series'] as List? ?? const [])
            .map((e) => UsageSeriesPoint.fromJson(e as Map<String, dynamic>))
            .toList(),
        fup: json['fup'] == null
            ? null
            : FupStatus.fromJson(json['fup'] as Map<String, dynamic>),
      );
}

/// One calendar day's traffic (bytes). Mirrors DailyUsagePoint from
/// schemas/usage.py (GET /me/usage-history). Dates are plain calendar days —
/// not localized, so month bucketing stays stable across timezones.
class DailyUsagePoint {
  DailyUsagePoint({
    required this.date,
    required this.uploadBytes,
    required this.downloadBytes,
    required this.totalBytes,
  });

  final DateTime date;
  final int uploadBytes;
  final int downloadBytes;
  final int totalBytes;

  factory DailyUsagePoint.fromJson(Map<String, dynamic> json) =>
      DailyUsagePoint(
        date: DateTime.parse(json['date'].toString()),
        uploadBytes: (json['upload_bytes'] as num?)?.toInt() ?? 0,
        downloadBytes: (json['download_bytes'] as num?)?.toInt() ?? 0,
        totalBytes: (json['total_bytes'] as num?)?.toInt() ?? 0,
      );
}

/// One calendar month's total traffic (bytes), aggregated client-side.
class MonthlyUsagePoint {
  MonthlyUsagePoint({required this.month, required this.bytes});

  /// First day of the month (day = 1).
  final DateTime month;
  final int bytes;
}

/// Long-history daily usage. Mirrors DailyUsageHistoryResponse from
/// schemas/usage.py (GET /me/usage-history). Sourced from the daily rollup
/// (Splynx backfill), reaching years further back than per-session accounting.
class UsageHistory {
  UsageHistory({
    required this.start,
    required this.end,
    required this.totalUploadBytes,
    required this.totalDownloadBytes,
    required this.totalBytes,
    this.points = const [],
  });

  final DateTime start;
  final DateTime end;
  final int totalUploadBytes;
  final int totalDownloadBytes;
  final int totalBytes;
  final List<DailyUsagePoint> points;

  factory UsageHistory.fromJson(Map<String, dynamic> json) => UsageHistory(
        start: DateTime.parse(json['start'].toString()),
        end: DateTime.parse(json['end'].toString()),
        totalUploadBytes: (json['total_upload_bytes'] as num?)?.toInt() ?? 0,
        totalDownloadBytes:
            (json['total_download_bytes'] as num?)?.toInt() ?? 0,
        totalBytes: (json['total_bytes'] as num?)?.toInt() ?? 0,
        points: (json['points'] as List? ?? const [])
            .map((e) => DailyUsagePoint.fromJson(e as Map<String, dynamic>))
            .toList(),
      );

  /// Aggregate the daily points into calendar-month totals, ascending. Gaps
  /// (months with no recorded usage) are simply absent — same as the daily
  /// source, which isn't zero-filled.
  List<MonthlyUsagePoint> toMonthly() {
    final byMonth = <String, int>{};
    for (final p in points) {
      final key = '${p.date.year}-${p.date.month.toString().padLeft(2, '0')}';
      byMonth[key] = (byMonth[key] ?? 0) + p.totalBytes;
    }
    final out = byMonth.entries.map((e) {
      final parts = e.key.split('-');
      return MonthlyUsagePoint(
        month: DateTime(int.parse(parts[0]), int.parse(parts[1])),
        bytes: e.value,
      );
    }).toList()
      ..sort((a, b) => a.month.compareTo(b.month));
    return out;
  }
}

DateTime? _toDate(dynamic v) {
  if (v == null) return null;
  return DateTime.tryParse(v.toString())?.toLocal();
}
